modules/tinymce/src/core/main/ts/init/LicenseKeyManager.ts

```diff
@@ -0,0 +1,133 @@
+
+const GplLicenseKeyManager: LicenseKeyManager = {
+  verify: () => Promise.resolve(true),
+  validate: (_editor, data) => {
+    const { plugin } = data || {};
+    // Premium plugins are not allowed if 'gpl' is given as the license_key
+    if (Type.isString(plugin)) {
+      return Promise.resolve(false);
+    } else {
+      return Promise.resolve(true);
+    }
+  },
+};
+
+const ADDON_KEY = 'commercial';
+const PLUGIN_CODE = 'licensekeymanager';
+
+const setup = (): LicenseKeyManagerLoader => {
+  const addOnManager = AddOnManager<LicenseKeyManager>();
+
+  // TODO: TINY-12081: Add integrity/checksum check
+  const add = async (addOn: LicenseKeyManagerAddon): Promise<void> => {
+    addOnManager.add(ADDON_KEY, addOn);
+  };
+
+  const load = (editor: Editor, suffix: string): void => {
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey !== 'gpl' && !Obj.has(addOnManager.urls, ADDON_KEY)) {
+      const url = `plugins/${PLUGIN_CODE}/plugin${suffix}.js`;
+
+      addOnManager.load(ADDON_KEY, url).catch(() => {
+        // eslint-disable-next-line no-console
+        console.error(`Failed to load license key manager from url ${url}`);
+        displayNotification(editor, 'Failed to load license key manager');
+      });
+    }
+  };
+
+  const isLoaded = (editor: Editor) => {
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey === 'gpl') {
+      return true;
+    } else {
+      return Type.isNonNullable(addOnManager.get(ADDON_KEY));
+    }
+  };
+
+  const init = (editor: Editor): void => {
+    const lockLicenseKeyManager = () => {
+      Object.defineProperty(editor, 'licenseKeyManager', {
+        writable: false,
+        configurable: false,
+        enumerable: true,
+      });
+    };
+
+    const licenseKey = Options.getLicenseKey(editor);
+    if (licenseKey === 'gpl') {
+      editor.licenseKeyManager = GplLicenseKeyManager;
+      lockLicenseKeyManager();
+      return;
+    }
+
+    // CommercialLicenseKeyManager should be nonnullable here as the
+    // editor will not load without a license key manager constructor
+    const CommercialLicenseKeyManager = addOnManager.get(ADDON_KEY);
+    if (Type.isNullable(CommercialLicenseKeyManager)) {
+      return;
+    }
+
+    const commercialLicenseKeyManagerApi = CommercialLicenseKeyManager(editor, addOnManager.urls[ADDON_KEY]);
+    editor.licenseKeyManager = commercialLicenseKeyManagerApi || {};
+    lockLicenseKeyManager();
+    const verify = editor.licenseKeyManager.verify;
+    const validate = editor.licenseKeyManager.validate;
+    if (!Type.isFunction(verify) || !Type.isFunction(validate)) {
+      // eslint-disable-next-line no-console
+      console.error(`Incorrect license key manager API`);
+      displayNotification(editor, 'Failed to intialize license key manager');
+      ForceReadonly.forceReadonly(editor);
+      return;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    verify(editor);
```

Comment:
How would this prevent loading. I would have assumed that this init would return a boolean promise or something is we are supposed to init the editor?

Reply:
The plan from early spiking stages was for verification/validation of the actual license key to be asynchronous and not block loading of the editor. If needed, the editor would be set to readonly after init. This was done as we have to consider retrieving cloud-based license keys as well.

Reply:
Right it might take a while to resolve. I guess that would work since the common case would be that they have access. 👍